### DISCUSSION TITLE
how do I set up next.config.js so that the class name only contains hash?

### DISCUSSION THREAD

**Discussion by Eugene0** [vercel/next.js#14156] [Help] [answered]
Hey, I use css modules in the project. node-sass and @zeit/next-sass I don't want to use. Can you please tell us how to configure next.config.js so that the class name only contains hash?
package.json:
```json
{
  "name": "learn-starter",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start"
  },
  "dependencies": {
    "next": "9.3.5",
    "react": "16.13.1",
    "react-dom": "16.13.1",
    "sass": "^1.26.8"
  }
}
```

![image (1)](https://user-images.githubusercontent.com/19935532/84580174-58f43b00-addd-11ea-90a1-9fd193da1d5d.png)

**Comment by illuminist** [comment]
In next.config.js, you need to change webpack config.
```js
{
  webpack: (config) => {
    const indexOfCSSLoaderRule = // You have to identify css-loader rules by yourself here
    config.module.rules[indexOfCSSLoaderRule].options.modules.localIdentName: '[hash:base64:5]'
  }
}
```

**Reply by Eugene0** [reply]
Thanks for the answer. Please tell me what should the variable indexOfCSSLoaderRule be equal to?

**Reply by Eugene0** [reply]
```node.js
module.exports = {
  webpack: (config) => {
    config.module.rules.push({
      test: /\.css$/i,
      loader: 'css-loader',
      options: {
        modules: {
          localIdentName: '[hash:base64:5]',
        },
      },
    })
    return config
  },
}
```

I tried this option. Unfortunately, it didn't work. Now instead of class names, it returns undefined.
![image](https://user-images.githubusercontent.com/19935532/84599852-c3f54e80-ae7d-11ea-86d7-6b5f3ec63745.png)

**Comment by mattcarlotta** [comment, marked as answer]
@Eugene0 Here's a working [repo](https://github.com/mattcarlotta/next-sass-localidentname) with a custom `localIdentName`.

**next.config.js**:
```js
const regexEqual = (x, y) =>
  x instanceof RegExp &&
  y instanceof RegExp &&
  x.source === y.source &&
  x.global === y.global &&
  x.ignoreCase === y.ignoreCase &&
  x.multiline === y.multiline;

module.exports = {
  webpack(config) {
    const sassRules = config.module.rules
      .find((rule) => typeof rule.oneOf === "object")
      .oneOf.find(
        (rule) =>
          rule.sideEffects === false &&
          regexEqual(rule.test, /\.module\.(scss|sass)$/)
      );

    sassRules.use = sassRules.use.map((rule) =>
      rule.loader.includes("css-loader/dist")
        ? {
            ...rule,
            options: {
              ...rule.options,
              modules: {
                ...rule.modules,
                localIdentName: "[hash:base64:5]",
              },
            },
          }
        : rule
    );

    return config;
  },
};
```

**Reply by Eugene0** [reply]
Wow, thanks for the repo. I launched it. I decided to look at how class names with a hash in develop will look. To see this, I changed a few lines:
```node.js
const regexEqual = (x, y) =>
  x instanceof RegExp &&
  y instanceof RegExp &&
  x.source === y.source &&
  x.global === y.global &&
  x.ignoreCase === y.ignoreCase &&
  x.multiline === y.multiline;

module.exports = {
  webpack(config) {
    const sassRules = config.module.rules
      .find((rule) => typeof rule.oneOf === "object")
      .oneOf.find(
        (rule) =>
          rule.sideEffects === false &&
          regexEqual(rule.test, /\.module\.(scss|sass)$/)
      );

      console.log(sassRules.use, ' sassRules.use')

    sassRules.use = sassRules.use.map((rule) => {
      return ({
        ...rule,
        options: {
          ...rule.options,
          modules: {
            ...rule.modules,
            localIdentName: "[hash:base64:5]",
          },
        },
      })    
    }      
    );

    return config;
  },
};
```

But unfortunately, I could not see it because an error appeared:

```
Failed to compile

./components/Example/Example.module.scss
ValidationError: Invalid options object. Style Loader has been initialized using an options object that does not match the API schema.
 - options has an unknown property 'modules'. These properties are valid:
   object { injectType?, attributes?, insert?, base? }
```


Maybe you know how to fix this?

**Reply by mattcarlotta** [reply]
You changed my config and added the `localIndentName` to all loaders, instead of just to the `css-loader`, which applies incorrect loader properties.

If you want to see what the classnames will look like in development, then open your dev console in the browser.

**Reply by Eugene0** [reply]
Thank. I opened the console in a browser. Class name contain component name. Now the class is 'Example_example__2hmjM' and I would like to get the class '2hmjM'

![image (1)](https://user-images.githubusercontent.com/19935532/84622945-dc4d8380-aee6-11ea-9e1b-4e6e9c30dc0f.png)

**Reply by mattcarlotta** [reply]
It's working as intended for me:
![](https://i.imgur.com/NMTfA3U.png)

**Comment by Eugene0** [comment]
I understood why it doesn’t work for me. 
```js
console.log(sassRules.use)
[
  {
    loader: 'C:\\Users\\User\\Desktop\\next-example2\\next-sass-localidentname\\node_modules\\css-loader\\dist\\cjs.js',
    options: {
      importLoaders: 3,
      sourceMap: true,
      onlyLocals: true,
      modules: [Object]
    }
  },
  {
    loader: 'C:\\Users\\User\\Desktop\\next-example2\\next-sass-localidentname\\node_modules\\next\\dist\\compiled\\postcss-loader\\index.js',
    options: { ident: '__nextjs_postcss', plugins: [Array], sourceMap: true }
  },
  {
    loader: 'C:\\Users\\User\\Desktop\\next-example2\\next-sass-localidentname\\node_modules\\resolve-url-loader\\index.js',
    options: { sourceMap: true }
  },
  {
    loader: 'C:\\Users\\User\\Desktop\\next-example2\\next-sass-localidentname\\node_modules\\sass-loader\\dist\\cjs.js',
    options: { sourceMap: true, sassOptions: {} }
  }
]
```
if you write like this, then it works for me 
```
rule.loader.includes("C:\\Users\\User\\Desktop\\next-example2\\next-sass-localidentname\\node_modules\\css-loader\\dist\\cjs.js")
```
![image (2)](https://user-images.githubusercontent.com/19935532/84625213-7f080100-aeeb-11ea-825b-9a4a31465dd4.png)



it works for me now, so thank you, it's really cool.
the line doesn't look very nice, maybe there's a way to fix it?
```
rule.loader.includes("C:\\Users\\User\\Desktop\\next-example2\\next-sass-localidentname\\node_modules\\css-loader\\dist\\cjs.js")
```

**Reply by mattcarlotta** [reply]
Looks like Windows uses back slashes for directories. Try this instead:
`rule.loader.includes("css-loader\\dist")`

**Reply by Eugene0** [reply]
thanks, this is a great solution

**Reply by LGmatrix13** [reply]
@Eugene0, is there an updated solution for next 11? I tried this and it doesn't seem to be working...

Thanks!

**Comment by Tylerian** [comment]
FYI today I made PR #14197 which should indirectly address this once merged.

**Reply by Eugene0** [reply]
thank you for telling me about it.

**Comment by herlon214** [comment]
Is there a solution for Next 12? I've been trying many different approaches and none of them are working.

**Comment by Sam-Kruglov** [comment]
try searching github code, I found this and works for me 
https://github.com/camomeui/camome/blob/562619d306a741c393b5098eb8f5df7e161b3698/apps/web/next.config.mjs#L26
uses https://github.com/camomeui/camome/blob/562619d306a741c393b5098eb8f5df7e161b3698/packages/utils/src/css/index.ts
